### PR TITLE
feat: add volume grid start lookup - ODD light part 2

### PR DESCRIPTION
**Pull request by asalzburger**
This PR adds the possibility to create a grid based start volume lookup for navigation. It reduces the time for 1000 lookups in the ODD light example from `20 ms` to `1.5 ms` and hence `1.5 µs` per lookup.

This sits on top of #2424 and will wait until this one is merged.

### CODE REVIEW

File: Core/include/Acts/Detector/IndexedRootVolumeFinderBuilder.hpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+namespace Acts {
+
+namespace Experimental {
+
+class DetectorVolume;
+
+/// @brief This is the interface for builders that create root volume finder
+/// delegates
+class IndexedRootVolumeFinderBuilder : public IRootVolumeFinderBuilder {
```
Comment (niermann999) [inline Core/include/Acts/Detector/IndexedRootVolumeFinderBuilder.hpp:27]: Can this be made final?

File: Core/include/Acts/Detector/IndexedRootVolumeFinderBuilder.hpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  /// The virtual interface definition for root volume finder builders
+  ///
+  /// @param gctx the geometry context at the creation of the internal structure
+  /// @param rootVolumes the root volumes to be used for the finder
+  ///
+  /// @return a shared detector object
+  DetectorVolumeUpdator construct(
+      const GeometryContext& gctx,
+      const std::vector<std::shared_ptr<DetectorVolume>>& rootVolumes)
+      const final;
```
Comment (niermann999) [inline Core/include/Acts/Detector/IndexedRootVolumeFinderBuilder.hpp:42]:
```suggestion
      const final override;
```

File: Core/include/Acts/Detector/detail/CylindricalDetectorHelper.hpp [hunk 3]
```diff
@@ -166,15 +168,82 @@ DetectorComponent::PortalContainer wrapInZR(
...
+  // The local logger
+  ACTS_LOCAL_LOGGER(getDefaultLogger("CylindricalDetectorHelper", logLevel));
+
+  ACTS_DEBUG("Estimate R/z/phi boundaries of  " << volumes.size()
+                                                << " volumes.");
+
+  // The return boundaries
+  std::array<std::vector<ActsScalar>, 3u> boundaries;
+
+  // The map for collecting
```
Comment (niermann999) [inline Core/include/Acts/Detector/detail/CylindricalDetectorHelper.hpp:191]:
```suggestion
  // The map for collecting shared volume boundary values
```

File: Core/include/Acts/Detector/detail/CylindricalDetectorHelper.hpp [hunk 4]
```diff
@@ -166,15 +168,82 @@ DetectorComponent::PortalContainer wrapInZR(
...
+  std::array<std::map<ActsScalar, size_t>, 3u> valueMaps;
+  auto& rMap = valueMaps[0u];
+  auto& zMap = valueMaps[1u];
+  auto& phiMap = valueMaps[2u];
+
+  auto fillMap = [&](std::map<ActsScalar, size_t>& map,
```
Comment (niermann999) [inline Core/include/Acts/Detector/detail/CylindricalDetectorHelper.hpp:197]:
```suggestion
  auto fillMap = [](std::map<ActsScalar, size_t>& map,
```

File: Core/include/Acts/Detector/detail/CylindricalDetectorHelper.hpp [hunk 5]
```diff
@@ -166,15 +168,82 @@ DetectorComponent::PortalContainer wrapInZR(
...
+      } else {
+        map[v] = 1u;
+      }
+    }
+  };
+
+  // Loop over the volumes and collect boundaries
+  for (const auto& v : volumes) {
+    if (v->volumeBounds().type() == VolumeBounds::BoundsType::eCylinder) {
+      auto bValues = v->volumeBounds().values();
```
Comment (niermann999) [inline Core/include/Acts/Detector/detail/CylindricalDetectorHelper.hpp:211]:
```suggestion
      const auto& bValues = v->volumeBounds().values();
```

File: Core/include/Acts/Detector/detail/CylindricalDetectorHelper.hpp [hunk 6]
```diff
@@ -166,15 +168,82 @@ DetectorComponent::PortalContainer wrapInZR(
...
+      ActsScalar phiMax = phiCenter + phiSector;
+      // Fill the maps
+      fillMap(rMap, {rMin, rMax});
+      fillMap(zMap, {zMin, zMax});
+      fillMap(phiMap, {phiMin, phiMax});
+    }
+  }
+
+  for (auto [im, map] : enumerate(valueMaps)) {
+    for (auto [key, value] : map) {
```
Comment (niermann999) [inline Core/include/Acts/Detector/detail/CylindricalDetectorHelper.hpp:234]: If the ```value``` is not needed anywhere, can't you fill the boundaries into the vector directly?

Comment (niermann999) [inline Core/include/Acts/Detector/detail/CylindricalDetectorHelper.hpp:234]:
```suggestion
  for (const auto [im, map] : enumerate(valueMaps)) {
    for (const auto [key, value] : map) {
```

File: Core/src/Detector/IndexedRootVolumeFinderBuilder.cpp [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+  for (const auto [ic0, c0] : Acts::enumerate(boundaries[0u])) {
+    if (ic0 > 0) {
+      Acts::ActsScalar v0 = 0.5 * (c0 + boundaries[0u][ic0 - 1]);
+      for (const auto [ic1, c1] : Acts::enumerate(boundaries[1u])) {
+        if (ic1 > 0) {
+          Acts::ActsScalar v1 = 0.5 * (c1 + boundaries[1u][ic1 - 1]);
+          if (casts ==
+              std::array<Acts::BinningValue, 2u>{Acts::binZ, Acts::binR}) {
+            Acts::Vector3 zrPosition{v1, 0., v0};
+            for (auto [iv, v] : Acts::enumerate(rootVolumes)) {
```
Comment (niermann999) [inline Core/src/Detector/IndexedRootVolumeFinderBuilder.cpp:36]:
```suggestion
            for (const auto [iv, v] : Acts::enumerate(rootVolumes)) {
```

File: Core/src/Detector/IndexedRootVolumeFinderBuilder.cpp [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    const std::array<std::vector<Acts::ActsScalar>, 2u>& boundaries,
+    const std::array<Acts::BinningValue, 2u>& casts) {
+  // Brute force loop over all bins & all volumes
+  for (const auto [ic0, c0] : Acts::enumerate(boundaries[0u])) {
+    if (ic0 > 0) {
+      Acts::ActsScalar v0 = 0.5 * (c0 + boundaries[0u][ic0 - 1]);
+      for (const auto [ic1, c1] : Acts::enumerate(boundaries[1u])) {
+        if (ic1 > 0) {
+          Acts::ActsScalar v1 = 0.5 * (c1 + boundaries[1u][ic1 - 1]);
+          if (casts ==
```
Comment (niermann999) [inline Core/src/Detector/IndexedRootVolumeFinderBuilder.cpp:33]: Can't you do that check once at the beginning, outside of the loop?

File: Core/src/Detector/IndexedRootVolumeFinderBuilder.cpp [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+#include "Acts/Detector/DetectorVolume.hpp"
+#include "Acts/Detector/detail/CylindricalDetectorHelper.hpp"
+#include "Acts/Detector/detail/GridAxisGenerators.hpp"
+#include "Acts/Navigation/DetectorVolumeFinders.hpp"
+#include "Acts/Utilities/Enumerate.hpp"
+
+namespace {
+
+template <typename Grid2D>
+void fillGridIndices2D(
```
Comment (niermann999) [inline Core/src/Detector/IndexedRootVolumeFinderBuilder.cpp:20]: Is this also expected to be used elsewhere or could this become a private helper method of the ```IndexedRootVolumeFinderBuilder```?

File: Core/src/Detector/IndexedRootVolumeFinderBuilder.cpp [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+Acts::Experimental::IndexedRootVolumeFinderBuilder::
+    IndexedRootVolumeFinderBuilder(std::vector<Acts::BinningValue> binning)
+    : m_casts(std::move(binning)) {
+  if (m_casts != std::vector<Acts::BinningValue>{Acts::binZ, Acts::binR}) {
+    throw std::invalid_argument("Online (z,r) binning is currently supported.");
+  }
+}
+
+Acts::Experimental::DetectorVolumeUpdator
+Acts::Experimental::IndexedRootVolumeFinderBuilder::construct(
```
Comment (niermann999) [inline Core/src/Detector/IndexedRootVolumeFinderBuilder.cpp:59]: Since this construct method is relying on some assumptions in its implementation (e.g. using ```CylindricalDetectorHelper```, passing on the first two cast values), it will only work for cylindrical volumes? Should it then not be named accordingly?

File: Examples/Algorithms/Geometry/include/ActsExamples/Geometry/VolumeAssociationTest.hpp [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "ActsExamples/Framework/IAlgorithm.hpp"
+#include "ActsExamples/Framework/ProcessCode.hpp"
+#include "ActsExamples/Framework/RandomNumbers.hpp"
+
```
Comment (niermann999) [inline Examples/Algorithms/Geometry/include/ActsExamples/Geometry/VolumeAssociationTest.hpp:14]:
```suggestion

#include <memory>
#include <string>
#include <vector>

```

File: Examples/Algorithms/Geometry/include/ActsExamples/Geometry/VolumeAssociationTest.hpp [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    std::size_t ntests = 1000;
+    /// The random number service
+    std::shared_ptr<const RandomNumbers> randomNumbers = nullptr;
+    /// The random number range
+    std::vector<Acts::ActsScalar> randomRange = {};
+    /// The detector
+    std::shared_ptr<const Acts::Experimental::Detector> detector = nullptr;
+  };
+
+  /// Construct the smearing algorithm.
```
Comment (niermann999) [inline Examples/Algorithms/Geometry/include/ActsExamples/Geometry/VolumeAssociationTest.hpp:43]:
```suggestion
  /// Construct the volume association test.
```

File: Examples/Algorithms/Geometry/include/ActsExamples/Geometry/VolumeAssociationTest.hpp [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+  ///
+  /// @param cfg is the algorithm configuration
+  /// @param level is the logging level
+  VolumeAssociationTest(const Config& cfg,
+                        Acts::Logging::Level level = Acts::Logging::INFO);
+
+  /// Build measurement from simulation hits at input.
```
Comment (niermann999) [inline Examples/Algorithms/Geometry/include/ActsExamples/Geometry/VolumeAssociationTest.hpp:50]: Can this comment be updated, too?

File: Examples/Scripts/Python/volume_association_test.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+    )
+    p.add_argument(
+        "-n", "--events", type=int, default=1000, help="Number of events to generate"
+    )
+    p.add_argument("-t", "--tests", type=int, default=10000, help="Tests per track")
+
+    args = p.parse_args()
+    geoContext = GeometryContext()
+
+    # Convert the detector surfaces to GDML
```
Comment (niermann999) [inline Examples/Scripts/Python/volume_association_test.py:61]:
```suggestion
    # Convert the detector surfaces from GDML
```

File: Tests/UnitTests/Plugins/Cuda/Seeding2/TestHostCuts.hpp [hunk 15]
```diff
@@ -46,6 +46,6 @@ class TestHostCuts : public Acts::IExperimentCuts<TestSpacePoint> {
   cutPerMiddleSP(
       std::vector<typename Acts::CandidatesForMiddleSp<
           const Acts::InternalSpacePoint<TestSpacePoint>>::value_type>
-          seedCandidates) const final override;
```
Comment (niermann999) [inline Tests/UnitTests/Plugins/Cuda/Seeding2/TestHostCuts.hpp:49]: Why was this removed, it does not seem to be related to the PR?

File: Core/include/Acts/Detector/IndexedRootVolumeFinderBuilder.hpp (hunk 1 above)
Comment (asalzburger) [reply Core/include/Acts/Detector/IndexedRootVolumeFinderBuilder.hpp:27]: It can and is being done.

File: Core/include/Acts/Detector/IndexedRootVolumeFinderBuilder.hpp (hunk 2 above)
Comment (asalzburger) [reply Core/include/Acts/Detector/IndexedRootVolumeFinderBuilder.hpp:42]: Somehow clang-tidy complained that one of the two is redundant?

File: Core/include/Acts/Detector/detail/CylindricalDetectorHelper.hpp (hunk 3 above)
Comment (asalzburger) [reply Core/include/Acts/Detector/detail/CylindricalDetectorHelper.hpp:191]: Added similar

File: Core/include/Acts/Detector/detail/CylindricalDetectorHelper.hpp (hunk 4 above)
Comment (asalzburger) [reply Core/include/Acts/Detector/detail/CylindricalDetectorHelper.hpp:197]: Done.

File: Core/include/Acts/Detector/detail/CylindricalDetectorHelper.hpp (hunk 6 above)
Comment (asalzburger) [reply Core/include/Acts/Detector/detail/CylindricalDetectorHelper.hpp:234]: Absolutely right, I think I wanted to add some statistics, but who needs to know that. Changed.

Comment (asalzburger) [reply Core/include/Acts/Detector/detail/CylindricalDetectorHelper.hpp:234]: Lines removed.

File: Core/src/Detector/IndexedRootVolumeFinderBuilder.cpp (hunk 7 above)
Comment (asalzburger) [reply Core/src/Detector/IndexedRootVolumeFinderBuilder.cpp:36]: Done.

File: Core/src/Detector/IndexedRootVolumeFinderBuilder.cpp (hunk 8 above)
Comment (asalzburger) [reply Core/src/Detector/IndexedRootVolumeFinderBuilder.cpp:33]: For now, yes, but I am planing to change the casting into compile time switches, then this will change anyway.

File: Core/src/Detector/IndexedRootVolumeFinderBuilder.cpp (hunk 9 above)
Comment (asalzburger) [reply Core/src/Detector/IndexedRootVolumeFinderBuilder.cpp:20]: The functionality will be expanded for xy grids, etc when we adapt this to telescope geometries, for the moment, I would leave it conventiently.

File: Core/src/Detector/IndexedRootVolumeFinderBuilder.cpp (hunk 10 above)
Comment (asalzburger) [reply Core/src/Detector/IndexedRootVolumeFinderBuilder.cpp:59]: See comment above.

File: Examples/Algorithms/Geometry/include/ActsExamples/Geometry/VolumeAssociationTest.hpp (hunk 11 above)
Comment (asalzburger) [reply Examples/Algorithms/Geometry/include/ActsExamples/Geometry/VolumeAssociationTest.hpp:14]: Updated.

File: Examples/Algorithms/Geometry/include/ActsExamples/Geometry/VolumeAssociationTest.hpp (hunk 12 above)
Comment (asalzburger) [reply Examples/Algorithms/Geometry/include/ActsExamples/Geometry/VolumeAssociationTest.hpp:43]: Updated.

File: Examples/Algorithms/Geometry/include/ActsExamples/Geometry/VolumeAssociationTest.hpp (hunk 13 above)
Comment (asalzburger) [reply Examples/Algorithms/Geometry/include/ActsExamples/Geometry/VolumeAssociationTest.hpp:50]: Changed, of course.

File: Tests/UnitTests/Plugins/Cuda/Seeding2/TestHostCuts.hpp (hunk 15 above)
Comment (asalzburger) [reply Tests/UnitTests/Plugins/Cuda/Seeding2/TestHostCuts.hpp:49]: clang-tidy suggestion (which I ran on the entire suite).

File: Examples/Scripts/Python/volume_association_test.py (hunk 14 above)
Comment (asalzburger) [reply Examples/Scripts/Python/volume_association_test.py:61]: Fixed.

File: Core/include/Acts/Detector/IndexedRootVolumeFinderBuilder.hpp (hunk 2 above)
Comment (niermann999) [reply Core/include/Acts/Detector/IndexedRootVolumeFinderBuilder.hpp:42]: Interesting, I guess a method cannot be final unless it already overrides? But since the entire class is now ```final```, I would actually replace ```final``` on the method with ```override```, if possible

File: Core/include/Acts/Detector/detail/CylindricalDetectorHelper.hpp [hunk 16]
```diff
@@ -166,15 +168,68 @@ DetectorComponent::PortalContainer wrapInZR(
...
+          bValues[CylinderVolumeBounds::BoundValues::eHalfLengthZ];
+      ActsScalar zMin = zCenter - zHalfLength;
+      ActsScalar zMax = zCenter + zHalfLength;
+      ActsScalar phiCenter =
+          bValues[CylinderVolumeBounds::BoundValues::eAveragePhi];
+      ActsScalar phiSector =
+          bValues[CylinderVolumeBounds::BoundValues::eHalfPhiSector];
+      ActsScalar phiMin = phiCenter - phiSector;
+      ActsScalar phiMax = phiCenter + phiSector;
+      // Fill the maps
```
Comment (niermann999) [inline Core/include/Acts/Detector/detail/CylindricalDetectorHelper.hpp:209]:
```suggestion
      // Fill the sets
```

**Review by niermann999 — APPROVED** [review]

File: Core/include/Acts/Detector/IndexedRootVolumeFinderBuilder.hpp (hunk 2 above)
Comment (paulgessinger) [reply Core/include/Acts/Detector/IndexedRootVolumeFinderBuilder.hpp:42]: Yeah, we follow one of the cppcore guidelines here, where `final` implies `override`, and both imply `virtual`.

**Review by niermann999 — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results
### :white_check_mark: All tests successful

|status|job|
|------|---|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32425743) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsWorkflow.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32425742) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32425741) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsCoreSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32425740) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32425739) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32425738) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateOrthogonalSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32425737) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateClusters.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32425736) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsPersistifyEDM.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32425735) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsKfRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32425734) |
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsExtrapolationAlgTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32425733) |
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsITkTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32425732) |
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32425731) |
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32425730) |
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32425729) |
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32425728) |
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32425727) |
